formulaic/materializers/narwhals.py

```diff
@@ -0,0 +1,216 @@
+                            spsparse.csc_matrix.multiply, solo_factors.values()
+                        )
+                    }
+                )
+            else:
+                factors.append(
+                    {
+                        ":".join(solo_factors): functools.reduce(
+                            numpy.multiply,
+                            (numpy.asanyarray(p) for p in solo_factors.values()),
+                        )
+                    }
+                )
+
+        for i, reversed_product in enumerate(
+            itertools.product(*(factor.items() for factor in reversed(factors)))
+        ):
+            if spec.output == "sparse":
+                out[names[i]] = scale * functools.reduce(
+                    spsparse.csc_matrix.multiply,
+                    (p[1] for p in reversed(reversed_product)),
+                )
+            else:
+                out[names[i]] = scale * functools.reduce(
+                    numpy.multiply,
+                    (numpy.array(p[1]) for p in reversed(reversed_product)),
+                )
+        return out
+
+    @override
+    def _combine_columns(
+        self, cols: Sequence[Tuple[str, Any]], spec: ModelSpec, drop_rows: Sequence[int]
+    ) -> pandas.DataFrame:
+        # If we are outputing a pandas DataFrame, explicitly override index
+        # in case transforms/etc have lost track of it.
+        pandas_index = None
+        if spec.output == "pandas" and isinstance(self.data, pandas.DataFrame):
+            pandas_index = cast(pandas.DataFrame, self.data).index
+            if drop_rows:
+                pandas_index = pandas_index.drop(
+                    cast(pandas.DataFrame, self.data_context).index[drop_rows]
+                )
+
+        # Special case no columns to empty csc_matrix, array, or DataFrame
+        if not cols:
+            values = numpy.empty((self.data.shape[0], 0))
+            if spec.output == "sparse":
+                return spsparse.csc_matrix(values)
+            if spec.output == "narwhals":
+                # TODO: Inconsistent with non-empty case below (where we use to-native)
+                return nw.from_native(values)
+            if spec.output == "numpy":
+                return values
+            return (
+                pandas.DataFrame(index=pandas_index)
+                if pandas_index is not None
+                else pandas.DataFrame(values)
+            )
+
+        # Otherwise, concatenate columns into model matrix
+        if spec.output == "sparse":
+            return spsparse.hstack([col[1] for col in cols])
+
+        # TODO: Can we do better than this? Having to reconstitute raw data
+        # does not seem ideal.
+        combined = nw.from_dict(
+            {
+                name: (nw.to_native(col) if isinstance(col, nw.Series) else col)
+                for name, col in cols
+            },
+            native_namespace=nw.get_native_namespace(self.__narwhals_data),
+        )
```

Comment:
You're right, this should be supported (and easier!) - I've opened an issue about it here https://github.com/narwhals-dev/narwhals/issues/756